components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,384 @@
+                             -> Result<Vec<PackageArchive>>
+        where T: UIWriter
+    {
+        let mut downloaded_artifacts = Vec::<PackageArchive>::new();
+
+        ui.status(Status::Downloading,
+                  format!("Downloading {} artifacts", expanded_idents.len()))?;
+
+        for ident in expanded_idents {
+            // TODO think through error handling here; failure to fetch, etc
+            // Probably worth keeping statistics
+            let archive: PackageArchive = self.get_cached_archive(ui, &ident)?;
+
+            downloaded_artifacts.push(archive);
+        }
+
+        Ok(downloaded_artifacts)
+    }
+
+    fn determine_latest_from_ident<T>(&self,
+                                      ui: &mut T,
+                                      ident: &PackageIdentTarget)
+                                      -> Result<Package>
+        where T: UIWriter
+    {
+        // Unlike in the install command, we always hit the online
+        // depot; our purpose is to sync with latest, and falling back
+        // to a local package would defeat that. Find the latest
+        // package in the proper channel from Builder API,
+        ui.status(Status::Determining,
+                  format!("latest version of {} in the '{}' channel",
+                          &ident, self.channel))?;
+        match self.fetch_latest_package_in_channel_for(ident, self.channel, self.token) {
+            Ok(latest_package) => {
+                ui.status(Status::Using,
+                          format!("{} as lastest matching {}", latest_package.ident, ident))?;
+                Ok(latest_package)
+            }
+            Err(Error::APIClient(APIError(StatusCode::NOT_FOUND, _))) => {
+                // In install we attempt to recommend a channel to look in. That's a bit of a
+                // heavyweight process, and probably a bad idea in the context of
+                // what's a normally a batch process. It might be ok to fall back to
+                // the stable channel, but for now, error.
+                ui.warn(format!("No releases of {} for exist in the '{}' channel",
+                                ident, self.channel))?;
+                Err(Error::PackageNotFound("".to_string()))
+            }
+            Err(e) => {
+                debug!("error fetching ident {}: {:?}", ident, e);
+                Err(e)
+            }
+        }
+    }
+
+    // This function and it's sibling get_cached_artifact in
+    // install.rs deserve to be refactored to eke out commonality.
+    /// This ensures the identified package is in the local cache,
+    /// verifies it, and returns a handle to the package's metadata.
+    fn get_cached_archive<T>(&self,
+                             ui: &mut T,
+                             package: &PackageIdentTarget)
+                             -> Result<PackageArchive>
+        where T: UIWriter
+    {
+        let fetch_artifact = || self.fetch_artifact(ui, package);
```

Comment:
This is always downloading the package first which seems incorrect.  I think we want to check to see if it is in the cache first.  If it is, we copy it over to the download path (or, do nothing if download path == cache path).